src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs

```diff
@@ -3499,44 +3500,37 @@ public sealed class SocketsHttpHandlerTest_ConnectCallback_Http2 : SocketsHttpHa
-                using (NamedPipeClientStream clientStream = new NamedPipeClientStream(".", pipeName: guid, PipeDirection.InOut, PipeOptions.WriteThrough | PipeOptions.Asynchronous, System.Security.Principal.TokenImpersonationLevel.Anonymous))
-                {
-                    await Task.WhenAll(serverStream.WaitForConnectionAsync(), clientStream.ConnectAsync());
-                    socketsHandler.ConnectCallback = (context, token) =>
-                    {
-                        return new ValueTask<Stream>(clientStream);
-                    };
+                await using NamedPipeClientStream clientStream = new(".", pipeName: guid, PipeDirection.InOut, PipeOptions.WriteThrough | PipeOptions.Asynchronous, TokenImpersonationLevel.Anonymous);
+                await clientStream.ConnectAsync(TestHelper.PassingTestTimeoutMilliseconds);
 
-                    using (HttpClient client = CreateHttpClient(handler))
-                    {
-                        client.DefaultVersionPolicy = HttpVersionPolicy.RequestVersionExact;
+                using HttpClientHandler handler = CreateHttpClientHandler();
+                using HttpClient client = CreateHttpClient(handler);
+                client.DefaultVersionPolicy = HttpVersionPolicy.RequestVersionExact;
 
-                        Task<string> clientTask = client.GetStringAsync($"{(options.UseSsl ? "https" : "http")}://{guid}/foo");
-                        await using (GenericLoopbackConnection loopbackConnection = await LoopbackServerFactory.CreateConnectionAsync(socket: null, serverStream, options))
-                        {
-                            await loopbackConnection.InitializeConnectionAsync();
+                GetUnderlyingSocketsHttpHandler(handler).ConnectCallback = (_, _) => new ValueTask<Stream>(clientStream);
 
-                            HttpRequestData requestData = await loopbackConnection.ReadRequestDataAsync();
-                            Assert.Equal("/foo", requestData.Path);
+                Assert.Equal("foo", await client.GetStringAsync($"{(useSsl ? "https" : "http")}://{guid}/foo"));
+            });
 
-                            await loopbackConnection.SendResponseAsync(content: "foo");
+            Task serverTask = Task.Run(async () =>
+            {
+                await using NamedPipeServerStream serverStream = new(pipeName: guid, PipeDirection.InOut, 1, PipeTransmissionMode.Byte, PipeOptions.WriteThrough | PipeOptions.Asynchronous);
+                await serverStream.WaitForConnectionAsync();
 
-                            string response = await clientTask;
-                            Assert.Equal("foo", response);
-                        }
-                    }
-                }
-            }
+                await using Http2LoopbackConnection loopbackConnection = await Http2LoopbackConnection.CreateAsync(socket: null, serverStream, new Http2Options { UseSsl = useSsl });
```

Comment:
should we use the Generic version so we have ability to test both Http 1.1 and 2?
I think adding the Dispose may be still worth or awe may add explicit method to disconnect.

Reply:
> HTTP/1.1 doesn't work over named pipes when using TLS 

Is this something we should track? cc @JamesNK in case it is a problem for https://github.com/dotnet/aspnetcore/issues/14207

Reply:
It does work if at least one side has pending reads to unblock the other.

If I'm looking at the [right part of Kestrel code](https://github.com/dotnet/aspnetcore/blob/main/src/Servers/Kestrel/Transport.NamedPipes/src/Internal/NamedPipeConnection.cs) for named pipes, they do have such logic (similar to their socket transport). So I'd imagine this should work with Kestrel.

Reply:
I would expect that the sever should always try to read the request, right? Perhaps is is shortcoming of out loopback code...?

Reply:
In this case, the server got stuck before even getting to the request. It was still trying to do the final write in the SslStream handshake.
So I guess in Kestrel's case it would depend on whether they're buffering the stream before/after SslStream.
If it's after they'd be stuck in the handshake the same way our loopback server would be. If it's before, it should be fine.